### DISCUSSION TITLE
More convenient access to extra asgi scope fields?

### DISCUSSION THREAD

**Discussion by hoodmane** [fastapi/fastapi#11219] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

There is a full gist with example code here, but see description first.
https://gist.github.com/hoodmane/d6f2b6fa3dfb79cafe958edfa3d6aab2

### Description

I am trying to get FastAPI working in cloudflare workers + Pyodide. I have [a custom asgi server](https://gist.github.com/hoodmane/d6f2b6fa3dfb79cafe958edfa3d6aab2#file-mini_asgi_server-py) to adapt between our runtime notion of `Request`/`Response` and asgi. The request handlers get extra context in an `environment` parameter which gives access to API keys and such. In this context, the `environment` parameter is very important so we'd like to make accessing it as simple as possible. From the point of view of an asgi server, the natural place to put it is in the `scope`. But accessing `scope` from fastapi isn't all that slick (though it's not terrible either).

We can do:
```py
from fastapi import FastAPI, Request

app = FastAPI()


@app.get("/hello")
async def root(req: Request, x: int = 0):
    env = req.scope["env"]
    return {"message": "Hello World", "SECRET_KEY": env.OPENAI_API_KEY}
```
but it would be nice to be able to say something like:
```py
from fastapi import FastAPI, Request
from mini_asgi_server import Environment

app = FastAPI()


@app.get("/hello")
async def root(env: Environment, x: int = 0):
    return {"message": "Hello World", "SECRET_KEY": env.OPENAI_API_KEY}
```
where somehow FastAPI could be told to get `Environment` type variables from `scope["env"]`. I looked at the code in `dependencies.utils.analyze_param` but I didn't find any code paths that look relevant for this. The handling of `Request` and friends looks quite hard-coded.

1. Is there some way to do this?
2. If not, would you accept a PR to add a way?

I was hoping for something like a pytest fixture where I can create a new one like:
```py
@pytest.fixture
def environment(request: Request) -> Environment:
   return request.scope["env"]
```

### Operating System

Other

### Operating System Details

Pyodide

### FastAPI Version

0.110.0

### Pydantic Version

2.6.3

### Python Version

4.12

**Comment by n8sty** [comment, marked as answer]
I think you can achieve this with a dependency:

```
def env(request: Request) -> Environment:
    return request.scope["env"]
```

Then in your route:

```
@app.get("/hello")
async def root(env: Environment = Depends(env), x: int = 0):
    return {"message": "Hello World", "SECRET_KEY": env.OPENAI_API_KEY}
```

**Reply by hoodmane** [reply]
Thanks @n8sty, that's quite helpful.
